### PR TITLE
Feat/editable profile

### LINKED ISSUE #127: [Feature Request]: Editable profile
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Feature Description

Profile is not editable in current situation as a result if user wants to change the email address or username it is not possible as a result in order to change any mistake while signing up is not correctable 

### Use Case

In this way if user make any mistake while signing up in writing email or username , user can correct that in profile page. That will enhance the usability of project

### Benefits

In this way if user make any mistake while signing up in writing email or username , user can correct that in profile page. That will enhance the usability of project

### Add ScreenShots

_No response_

### Priority

High

### Record

- [X] I have read the Contributing Guidelines
- [X] I'm a GSSOC'24 contributor
- [X] I have starred the repository

### PR DESCRIPTION
**Pull request by anwesha2002**
# Pull Request

### Title
Feature update :  Editable profile #127

### Description
this PR adds the feature of the editable profile. User can edit their profile in profile section

### Related Issues
Fixes #127 

### Changes Made
- An edit form added for user profile update.
- profile_update route and controller is added in the backend
- All the necessary changes made in the frontend and the backend for this feature
- Zod verification is also taken care of

### Checklist 

- [x ] I have tested the changes locally
- [x ] Documentation has been updated (if necessary)
- [x ] Changes are backward-compatible

### Screenshots (if applicable)
https://github.com/VaibhavArora314/StyleShare/assets/116761608/8982eb69-0655-4e2e-84bf-a87278fbb958

### Additional Notes
HMU if any changes needed

Footer

### CODE REVIEW

**Review by VaibhavArora314 — CHANGES_REQUESTED** [review]
image
resolve these errors

**Review by VaibhavArora314 — CHANGES_REQUESTED** [review]
looks good to me, some things to change:
1. don't use contains in updating as it is affecting all users
2. check if email already in use by other user
3. if user changes email, mark user verified status to false

**Comment by anwesha2002** [comment]
@VaibhavArora314 I think there is no function named 'updateFirst'. in order to do so first will have to do findFirst and then 'update' that record. otherwise if i use updateMany, all the data contains that userID will be updated means in follow list, or in post or where ever it is in use.
so should I change it??

**Comment by VaibhavArora314** [comment]
> @VaibhavArora314 I think there is no function named 'updateFirst'. in order to do so first will have to do findFirst and then 'update' that record. otherwise if i use updateMany, all the data contains that userID will be updated means in follow list, or in post or where ever it is in use. so should I change it??

there is a method named update

**Comment by VaibhavArora314** [comment]
or you can dont use contains while updating since that is affecting multiple user I guess

**Comment by anwesha2002** [comment]
> don't use contains while updating since that is affecting multiple users. you can use updateMany but don't use contains for id in where

okay

File: frontend/src/components/profileForm.tsx [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                toast.error("Email cannot be empty")
+            }
+            if(!username){
+                toast.error("Username cannot be empty")
+            }
+            const response = await axios.put(`/api/v1/user/update/${user?.id}`,updatedUser);
+            toast.success(response.data.message)
+            dismiss()
+        }catch (e){
+            const axiosError = e as AxiosError<{
```
Comment (akbatra567) [inline frontend/src/components/profileForm.tsx:38]: Avoid `as` whenever possible as it does a force conversion to the specified type. 

Potential solution is: 
```suggestion
            const axiosError: AxiosError<{ error: { message: string; }}> = e;
```
You would need to test if it directly works also if you fell the error is commonly used, you can create a type and pass directly. 

For instance: 
```ts
type ErrorType = { error: { message: string; }};
const axiosError: AxiosError<ErrorType> = e;
```

File: frontend/src/components/profileForm.tsx [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                error: {
+                    message: string;
+                };
+            }>;
+
+            setError((e) => {
+                if (axiosError?.response?.data?.error)
+                    return axiosError?.response?.data?.error as typeof e;
```
Comment (akbatra567) [inline frontend/src/components/profileForm.tsx:46]: See here also if you can avoid type convertion using `as` keyword. If it doesn't work you can avoid. Do try and let me know.

**Review by Ultimateutkarsh11 — CHANGES_REQUESTED** [review]
@anwesha2002 LGTM

**Review by VaibhavArora314 — CHANGES_REQUESTED** [review]
looking but after the user updates his details, the updated details should be shown on the profile as currently, it is showing old details

**Comment by anwesha2002** [comment]
> looking good but after the user updates his details, the updated details should be shown on the profile as currently, it is showing old details and user has to reload to see updated details

is it okay now??

**Review by VaibhavArora314 — APPROVED** [review]

**Review by Ultimateutkarsh11 — APPROVED** [review]

**Review by akbatra567 — APPROVED** [review]

**Pull request merged**